### PR TITLE
darwin.apple_sdk: update to 10.13

### PR DESCRIPTION
**Pull request by uri-canva**
###### Motivation for this change

10.10 SDKs don't support type arguments in Foundation classes, leading to "type arguments cannot be applied to non-parameterized class" errors.

Based on 377cef8d16c47df74d2653432d9bba968236c8a0.

## WARNING:

This will cause mass rebuilds on Darwin, and should be merged into staging I think, rather than master, please advise whether I should submit the PR against staging instead (it has merge conflicts between master and staging so if I submit it against staging it won't apply cleanly on master and vice versa).

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by uri-canva** [comment]
cc @copumpkin

**Comment by LnL7** [comment]
Only updating the sdk isn't an option because CoreFoundation is still at 10.10, but @copumpkin has made some pretty exciting progress there recently.

**Comment by uri-canva** [comment]
I haven't figured out exactly how it all connects together, but there's two different CoreFoundation right? An impure one and a pure one. I assume the impure one is this one, from the SDKs, and the pure one is one built from source? Are you saying we can't update the frameworks because they'd use the 10.10 CoreFoundation that is still from source?

**Comment by copumpkin** [comment]
@uri-canva thanks for the contribution! I'd take a look at https://github.com/NixOS/nixpkgs/pull/37403 since @matthewbauer also tried something similar. Since then he's added a package for `pbzx` so that at least can be less hackish. But I'd also hold off on this for a couple of reasons:

1. As @LnL7 said, the pure `CoreFoundation` we have is going to break with this. I just opened #46704  which should improve that but I'd rather hold off on changing too much at once so we can attribute weird failures precisely. Once that has stabilized, I definitely want to update the SDK and several other Apple packages
2. I'm not 100% sure 10.13 is a good version for this. The reason is that we do try to maintain some backwards-compatibility for folks who aren't on the latest version. I think our current bound goes back to 10.11. It's quite possible that 10.13 will work just fine on older systems, but I'd want to make sure before merging.

In short, changes like this have far-reaching implications and touch a bajillion packages, so we have a fairly high bar for merging them. So just to set your expectations, I wouldn't expect this to get merged in the next few weeks (because of 1 above, and because I'll be on vacation for a couple of weeks starting next Saturday) unless one of the Darwin maintainers suddenly has far more free time than we normally do 😄

**Comment by copumpkin** [comment]
Also, to answer your specific question: yes, there are two `CoreFoundation` packages and this SDK defines one of them. Basically, these SDKs have dubious licensing (even just for headers) and we'd rather not have our entire stdenv depend on them, since Hydra refuses to build things like that. Note that we're currently somewhat lax around this but making the entire stdenv depend on it would be a big step in the wrong direction IMO. At the same time, the publicly available `CoreFoundation` doesn't include all the stuff we need, so we do (occasionally) need to refer to the SDK version. To further complicate matters, you get all sorts of weird fun when you link against both versions of `CoreFoundation`, often transitively, so we have workarounds for that. It's been a constant pain in the ass, and it's mostly under control right now apart from forcing us to stay on an old SDK. The PR I just opened should allow us to finally advance but has changed lots of stuff so I want to keep an eye on it for a while.

**Comment by matthewbauer** [comment]
While it may be true that newer SDK will break things, it is usually really obvious looking at Hydra. Making breakages reproducible means we can just revert the bad commit. This is one of my favorite things about Nix!

I would shoot for using 10.13 SDK TBH. Apple is usually pretty good at not breaking old code so I think it should be okay. I know of a few things introduced in SDK that is already being used in the wild. Plus if anyone wants to use older macOS versions, they can just use the 18.09 or 18.03 channel.

**Comment by uri-canva** [comment]
Thanks for the clarifications @copumpkin, that helped.

I was also a bit uncertain whether it would be better to update the stack from the bottom up when I added `#define TARGET_OS_OSX 1` to `Libsystem` even though the source we build it from doesn't have it, because the 10.13 SDKs require it.

I will work around the need for an updated Apple SDKs in #46700, so it's not blocked.

> I'm not 100% sure 10.13 is a good version for this. The reason is that we do try to maintain some backwards-compatibility for folks who aren't on the latest version.

This is related to the impure CoreFoundation right? Because if the whole stack was pure it wouldn't matter at all what version the host macOS was, since everything would be built by Nix? Or is this about source compatibility of the derivations?

**Comment by lionello** [comment]
Was hoping that this patch would fix issues I'm having with `coreutils` by being able to build coreutils against libsystem ≥10.13, but it appears that it's still using the old 10.11.6. Any idea why?

**Comment by copumpkin** [comment]
@lionello because that stuff is governed by a completely separate set of packages in apple-source-releases. This is only used as a last resort. The apple-source-releases thing was the next thing I was going to look at, although I'm going on vacation soon so might not get to it for a bit.

Out of curiosity, what's missing in 10.11.6 that affects coreutils? We've been intentionally hiding some comparatively new symbols for libsystem in particular because we want to remain compatible with the last 3 or so major macOS revisions, so even if we update libsystem you might not get what you want for a while.

**Comment by lionello** [comment]
> Out of curiosity, what's missing in 10.11.6 that affects coreutils? We've been intentionally hiding some comparatively new symbols for libsystem in particular because we want to remain compatible with the last 3 or so major macOS revisions, so even if we update libsystem you might not get what you want for a while.

`futimens`, without it `touch` (and other utils) only handle µs resolution file times, even though the underlying FS likely had ns resolution. This means that created files never really have the exact expected time. Since libsystem 10.11.6 lacks `futimens`, coreutils falls back to `futimes`.

**Comment by uri-canva** [comment]
Thanks again for all the context, it's been very helpful.
I will close this PR since I don't plan on working it for now but I'll keep an eye out on developments from others.
